File: frontend/src/api/metrics.ts
import api from '@/api'
import type { ConnectionStates, MetricDataPoint, MetricSummary } from '@/types/metrics'

export async function getMetricNames(): Promise<string[]> {
  const res = await api.get('/metrics/names')
  return res.data.data as string[]
}

export async function getMetricTimeSeries(
  metricName: string,
  since = '1h',
  interval = '5m',
  label?: string,
): Promise<MetricDataPoint[]> {
  const params: Record<string, string> = { since, interval }
  if (label) params.label = label
  const res = await api.get(`/metrics/${metricName}`, { params })
  return res.data.data as MetricDataPoint[]
}

export interface MetricSummaryResult {
  summary: MetricSummary
  connectionStates: ConnectionStates
}

export async function getMetricSummary(): Promise<MetricSummaryResult> {
  const res = await api.get('/metrics/summary')
  return {
    summary: res.data.data as MetricSummary,
    connectionStates: (res.data.connection_states ?? {}) as ConnectionStates,
  }
}

// ── Route Stats ──
// 后端按 stats_type 返回不同字段组合（qps/bandwidth/error_rate/latency），
// 统一为扁平可选字段结构，与 app/services/metrics_service.py 的返回保持一致
export interface RouteStatsItem {
  route_id: string
  uri?: string
  // qps
  requests_per_sec?: number
  total_requests?: number
  sample_count?: number
  // bandwidth
  direction?: string
  bytes_per_sec?: number
  total_bytes?: number
  // error_rate
  client_errors?: number
  server_errors?: number
  error_rate_pct?: number
  // latency
  latency_type?: string
  avg_latency_ms?: number
  max_latency_ms?: number
}

export async function getRouteStats(
  statsType: string,
  since: string,
  limit: number,
  latencyType?: string,
): Promise<RouteStatsItem[]> {
  const params: Record<string, string | number> = { stats_type: statsType, since, limit }
  if (latencyType) params.latency_type = latencyType
  const res = await api.get('/metrics/route-stats', { params })
  return res.data.data as RouteStatsItem[]
}

// ── Status Analysis ──
export interface StatusAnalysisItem {
  status_class: string
  request_count: number
  percentage: number
}

export async function getStatusAnalysis(since: string): Promise<StatusAnalysisItem[]> {
  const res = await api.get('/metrics/status-analysis', { params: { since } })
  return res.data.data as StatusAnalysisItem[]
}

// ── Time Comparison ──
export interface DayOverDayData {
  today_requests: number
  yesterday_requests: number
  change_rate: number
  data_quality: string
}

export interface HourlyDistributionItem {
  hour_of_day: number
  day_of_week: number
  request_count: number
}

export async function getTimeComparison(
  comparisonType: string,
  days?: number,
): Promise<DayOverDayData | HourlyDistributionItem[]> {
  const params: Record<string, string | number> = { comparison_type: comparisonType }
  if (days) params.days = days
  const res = await api.get('/metrics/time-comparison', { params })
  return res.data.data as DayOverDayData | HourlyDistributionItem[]
}

// ── Node Health ──
export interface NodeHealthItem {
  node_ip: string
  status?: number
  last_seen?: string
  dict_name?: string
  capacity_bytes?: number
  free_bytes?: number
  usage_percent?: number
}

export async function getNodeHealth(
  healthType: string,
  statusFilter?: string,
): Promise<NodeHealthItem[]> {
  const params: Record<string, string> = { health_type: healthType }
  if (statusFilter) params.status = statusFilter
  const res = await api.get('/metrics/node-health', { params })
  return res.data.data as NodeHealthItem[]
}

export async function getRouteNameMap(edgeUuids: string[]): Promise<Record<string, { name: string; uri: string }>> {
  if (edgeUuids.length === 0) return {}
  const res = await api.post('/routes/by-edge-uuids', edgeUuids)
  const routes = res.data.items || []
  const map: Record<string, { name: string; uri: string }> = {}
  routes.forEach((r: { edge_uuid: string; name: string; uri: string }) => {
    map[r.edge_uuid] = { name: r.name, uri: r.uri }
  })
  return map
}

File: backend/app/api/v1/metrics.py
"""Metrics query API endpoints.

Provides read-only access to Edge node metrics stored in ClickHouse via
the esapm OpenTelemetry pipeline.

"""

from fastapi import APIRouter, HTTPException, Query

from app.services.metrics_service import (
    query_metric_names,
    query_time_series,
    query_summary,
    query_route_stats,
    query_status_analysis,
    query_time_comparison,
    query_node_health,
)

router = APIRouter(tags=["metrics"])

VALID_ROUTE_STATS_TYPES = {"qps", "bandwidth", "error_rate", "latency"}

@router.get("/metrics/names")
async def get_metric_names():
    names = query_metric_names()
    return {"data": names}


@router.get("/metrics/summary")
async def get_metrics_summary():
    data = query_summary()
    return {"data": data}


@router.get("/metrics/route-stats")
async def get_route_stats(
    stats_type: str = Query("qps"),
    since: str = Query("24h", pattern=r"^\d+[smhd]$"),
    limit: int = Query(10, ge=1, le=100),
    latency_type: str = Query("request"),
):
    if stats_type not in VALID_ROUTE_STATS_TYPES:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid type. Valid: {', '.join(sorted(VALID_ROUTE_STATS_TYPES))}"
        )
    data = query_route_stats(
        stats_type=stats_type,
        since=since,
        limit=limit,
        latency_type=latency_type,
    )
    return {"data": data}


@router.get("/metrics/status-analysis")
async def get_status_analysis(
    since: str = Query("24h", pattern=r"^\d+[smhd]$"),
):
    data = query_status_analysis(since=since)
    return {"data": data}


VALID_COMPARISON_TYPES = {"day_over_day", "hourly_distribution", "week_over_week"}


@router.get("/metrics/time-comparison")
async def get_time_comparison(
    comparison_type: str = Query("day_over_day"),
    days: int = Query(7, ge=1, le=30),
):
    if comparison_type not in VALID_COMPARISON_TYPES:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid type. Valid: {', '.join(sorted(VALID_COMPARISON_TYPES))}"
        )
    data = query_time_comparison(
        comparison_type=comparison_type,
        days=days,
    )
    return {"data": data}


VALID_HEALTH_TYPES = {"status", "resource"}


@router.get("/metrics/node-health")
async def get_node_health(
    health_type: str = Query("status"),
    status: str | None = Query(None),
):
    if health_type not in VALID_HEALTH_TYPES:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid type. Valid: {', '.join(sorted(VALID_HEALTH_TYPES))}"
        )
    data = query_node_health(
        health_type=health_type,
        status_filter=status,
    )
    return {"data": data}


@router.get("/metrics/{metric_name}")
async def get_metric_time_series(
    metric_name: str,
    since: str = Query("1h", pattern=r"^\d+[smhd]$"),
    interval: str = Query("5m", pattern=r"^\d+[sm]$"),
    label: str | None = Query(None),
):
    data = query_time_series(
        metric_name=metric_name,
        since=since,
        interval=interval,
        label=label,
    )
    return {"data": data}
